### DISCUSSION TITLE
page numbering for dompdf when content exceeds 1 page;

### DISCUSSION THREAD

**Discussion by solmazb128** [dompdf/dompdf#3541] [Project Development]
Hi all;
consider i use following to create html content for dompdf in codeigniter 3 controller method:

```
public function generatePageHtml($report_name, $title1, $title2 = null, $title3 = null, $content) { 
        $html .= '<div style="position: relative; z-index: 2; padding: 20px;">'; 
        $html .= '<div style="margin: 10px 25px 4px 5px;">'; 
        $html .= $content;
        $html .= '</div>';
        $html .= '<div style="position: absolute; bottom: 20px; left: 20px; font-size: 12px; font-weight: bold; font-style: italic; color: black;"> ' . $report_name .' - '. $title1 .' - '. $title2 . '</div>';
        $html .= '</div>'; 
        $html .= '</div>'; 
        return $html;
    }
```
    
and i use that in a for loop for each content fetched from model which includes variable columns of several sql tables:

```
foreach ($value as $sub_item) {
         $html_content .= $this->generatePageHtml($report_name,$translated_section, $translated_subtitle, null,$sub_item);
 }
```
i use convas as follow for page numbering:

  ```
      $canvas = $dompdf->getCanvas();
        $font = $dompdf->getFontMetrics()->getFont('times', 'normal');
        $size = 10;
        $page_number_text = "{PAGE_NUM}";
        $canvas->page_text(
            820, 
            570, 
            $page_number_text, 
            $font,
            $size,
            array(0, 0, 0), 
            null, 
            null 
        );
```
in final, i should add a table of contents and related page numbers to generated pdf;
here some content could exceeds one pdf page (A4 Landscape); thus page numbering should consider that;
numbering is truely applyed on pdf; 
but as some contents exceeds one pdf page, the usual numbering failed considering that in true page numbering;
how should i extract title 1, title 2 and its page number to create TOC for pdf;
perhaps, if i could define each time what is the last page number when i run generatePageHtml, so i could be able to add numbering to TOC truely.

**Comment by bsweeney** [comment]
Right now, the only way to do this would be to perform some content tracking and PDF source manipulation at runtime. I described a mechanism you could use to achieve this here: https://github.com/dompdf/dompdf/issues/1756#issuecomment-1006718311

I would write this differently now, using callbacks instead of inline script. But the overall mechanism doesn't change.
